src/modules/muorb/slpi/uORBProtobufChannel.cpp

```diff
@@ -44,29 +46,121 @@ static MUORBTestType test_to_run;
+{
+	PX4_INFO("Subscribing to %s on remote side", messageName);
+
+	if (muorb_func_ptrs.subscribe_func_ptr) {
+		pthread_mutex_lock(&_tx_mutex);
+		int16_t rc = muorb_func_ptrs.subscribe_func_ptr(messageName);
+		pthread_mutex_unlock(&_tx_mutex);
+		return rc;
+	}
+
+	PX4_ERR("subscribe_func_ptr is null in %s", __FUNCTION__);
+	return -1;
+}
+
+int16_t uORB::ProtobufChannel::remove_subscription(const char *messageName)
+{
+	PX4_INFO("Unsubscribing from %s on remote side", messageName);
+
+	if (muorb_func_ptrs.unsubscribe_func_ptr) {
+		pthread_mutex_lock(&_tx_mutex);
+		int16_t rc = muorb_func_ptrs.unsubscribe_func_ptr(messageName);
+		pthread_mutex_unlock(&_tx_mutex);
+		return rc;
+	}
+
+	PX4_ERR("unsubscribe_func_ptr is null in %s", __FUNCTION__);
+	return -1;
+}
+
+int16_t uORB::ProtobufChannel::register_handler(uORBCommunicator::IChannelRxHandler *handler)
+{
+	_RxHandler = handler;
+	return 0;
+}
+
+int16_t uORB::ProtobufChannel::send_message(const char *messageName, int32_t length, uint8_t *data)
+{
+	bool is_not_slpi_log = true;
+
+	if ((strcmp(messageName, "slpi_debug") == 0) || (strcmp(messageName, "slpi_error") == 0)
+	    || (strcmp(messageName, "muorb_test") == 0)) { is_not_slpi_log = false; }
+
+	if (muorb_func_ptrs.topic_data_func_ptr) {
+		PX4_INFO("Got message for topic %s", messageName);
+		std::string temp(messageName);
+		int has_subscribers = 0;
+		pthread_mutex_lock(&_rx_mutex);
+		has_subscribers = _AppsSubscriberCache[temp];
+		pthread_mutex_unlock(&_rx_mutex);
+
+		if ((has_subscribers) || (is_not_slpi_log == false)) {
+			PX4_INFO("Sending message for topic %s", messageName);
+			pthread_mutex_lock(&_tx_mutex);
+			int16_t rc = muorb_func_ptrs.topic_data_func_ptr(messageName, data, length);
+			pthread_mutex_unlock(&_tx_mutex);
+			return rc;
+		}
+
+		PX4_INFO("Skipping message for topic %s", messageName);
+		return 0;
+	}
+
+	if (is_not_slpi_log) { PX4_ERR("topic_data_func_ptr is null in %s", __FUNCTION__); }
+
+	return -1;
+}
+
 static void *test_runner(void *)
 {
 	PX4_INFO("test_runner called");
+	uORB::ProtobufChannel *channel = uORB::ProtobufChannel::GetInstance();
 
 	switch (test_to_run) {
 	case ADVERTISE_TEST_TYPE:
-		(void) muorb_func_ptrs.advertise_func_ptr(muorb_test_topic_name);
+		(void) channel->topic_advertised(muorb_test_topic_name);
 		break;
 
 	case SUBSCRIBE_TEST_TYPE:
-		(void) muorb_func_ptrs.subscribe_func_ptr(muorb_test_topic_name);
+		(void) channel->add_subscription(muorb_test_topic_name, 1);
 		break;
 
 	case UNSUBSCRIBE_TEST_TYPE:
-		(void) muorb_func_ptrs.unsubscribe_func_ptr(muorb_test_topic_name);
+		(void) channel->remove_subscription(muorb_test_topic_name);
 		break;
 
 	case TOPIC_TEST_TYPE: {
 			uint8_t data[MUORB_TEST_DATA_LEN];
 
 			for (uint8_t i = 0; i < MUORB_TEST_DATA_LEN; i++) { data[i] = i; }
 
-			(void) muorb_func_ptrs.topic_data_func_ptr(muorb_test_topic_name, data, MUORB_TEST_DATA_LEN);
```

Comment:
Leave this as direct call to muorb_func_ptrs so that we can skip the checks for testing in the send_message function.